src/memray/_memray/source.cpp

```diff
@@ -25,22 +25,32 @@ namespace memray::io {
 FileSource::FileSource(const std::string& file_name)
 : d_file_name(file_name)
 {
-    d_stream.open(d_file_name, std::ios::binary | std::ios::in);
-    if (!d_stream) {
+    d_raw_stream.open(d_file_name, std::ios::binary | std::ios::in);
+    if (!d_raw_stream) {
         throw IoError{"Could not open file " + file_name + ": " + std::string(strerror(errno))};
     }
+    uint32_t header;
+    d_raw_stream.read(reinterpret_cast<char*>(&header), 4);
+    d_raw_stream.seekg(0, std::ios::beg);
+
+    if (header == 0x184D2204) {
+        d_stream = std::make_unique<lz4_stream::istream>(d_raw_stream);
+    } else {
+        d_stream = std::make_unique<std::ifstream>();
+        dynamic_cast<std::ifstream*>(d_stream.get())->open(d_file_name, std::ios::binary | std::ios::in);
```

Comment:
There's a TOCTOU bug here. You're reopening the file after checking the magic, by which point it could have been rewritten with a different header.